File: python/src/addTwoNumbers.py
# https://leetcode.com/problems/add-two-numbers


from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        dummy = ListNode()
        tail = dummy

        carry = 0
        while l1 or l2 or carry:
            node = ListNode()
            tail.next = node
            tail = tail.next

            n1 = l1.val if l1 else 0
            n2 = l2.val if l2 else 0
            carry, node.val = divmod(n1 + n2 + carry, 10)

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        return dummy.next

File: go/addTwoNumbers.go
// https://leetcode.com/problems/add-two-numbers

package leetcode

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{}
	tail := dummy
	carry := 0

	for l1 != nil || l2 != nil || carry != 0 {
		node := &ListNode{}
		tail.Next = node
		tail = tail.Next

		n1 := 0
		if l1 != nil {
			n1 = l1.Val
		}

		n2 := 0
		if l2 != nil {
			n2 = l2.Val
		}

		digit := n1 + n2 + carry
		node.Val = digit % 10
		carry = digit / 10

		if l1 != nil {
			l1 = l1.Next
		}

		if l2 != nil {
			l2 = l2.Next
		}
	}

	return dummy.Next
}
